File: Arhitecture/Lab3/ElfParser.cpp
//
// Created by mike on 12/23/22.
//

#include <sstream>
#include <algorithm>
#include <string>
#include "command_parser.h"
#include "ElfParser.h"

namespace parser {
    namespace {
        std::string decimal_to_binary(int x) {
            std::stringstream ss;
            for (int i = 0; i < 8; ++i) {
                ss << (x % 2);
                x /= 2;
            }
            return ss.str();
        }
    }

    int ElfParser::cnt(int offset, int num) const {
        int ans = 0;
        for (int i = num - 1; i >= 0; --i) {
            ans = ans * 256 + bytes[offset + i];
        }
        return ans;
    }

    void ElfParser::parseHeader() {
        int shoff = cnt(32, 4);
        int shentsiz = cnt(46, 2);
        int shnum = cnt(48, 2);
        int shstrndx = cnt(50, 2);
        int go = shoff + shentsiz * shstrndx;
        int offset12 = cnt(go + 16, 4);
        for (int j = 0, i = shoff; j < shnum; ++j, i += 40) {
            std::string name = GetName(offset12 + cnt(i, 4));
            int params[10];
            for (int k = 0; k < 10; ++k) {
                params[k] = cnt(i + 4 * k, 4);
            }
            sections[name] = std::make_shared<utility::section>(params[0], params[1],
                                                                params[2], params[3],
                                                                params[4], params[5],
                                                                params[6], params[7],
                                                                params[8], params[9]);
        }
    }

    std::string ElfParser::GetName(int offset) const {
        std::stringstream ss;
        while (bytes[offset] != 0) {
            ss << (char) bytes[offset++];
        }
        return ss.str();
    }

    void ElfParser::parseText() {
        std::shared_ptr<utility::section> text = sections[".text"];
        writer->writeLine(".text");
        int offset = text->offset;
        int size = text->size;
        int addr = text->addr;
        for (int i = 0; i < size; i += 4) {
            std::stringstream ss;
            std::string byte1 = decimal_to_binary(bytes[offset]);
            std::reverse(byte1.begin(), byte1.end());
            std::string byte2 = decimal_to_binary(bytes[offset + 1]);
            std::reverse(byte2.begin(), byte2.end());
            std::string byte3 = decimal_to_binary(bytes[offset + 2]);
            std::reverse(byte3.begin(), byte3.end());
            std::string byte4 = decimal_to_binary(bytes[offset + 3]);
            std::reverse(byte4.begin(), byte4.end());
            ss << byte4 << byte3 << byte2 << byte1;
            std::string command = CommandParser::GetRiscvCommand(addr + i, ss.str());
            std::string label;
            if (symTable.count(addr + i) && elements[symTable[addr + i]]->type == "FUNC") {
                label = elements[symTable[addr + i]]->name;
            }
            char* buffer = new char[1024];
            if (!label.empty()) {
                sprintf(buffer, "%08x   <%s>:", addr + i, label.c_str());
                writer->writeLine(buffer);
            }
            delete[] buffer;
            writer->write(command);
            offset += 4;
        }
        writer->writeLine("");
    }

    void ElfParser::parseSymTab() {
        auto symtab = sections[".symtab"];
        int offset = symtab->offset;
        int num = symtab->size / 16;
        for (int i = 0; i < num; ++i) {
            std::stringstream ss;
            for (int j = 0; j < 4; ++j) {
                std::string str = decimal_to_binary(bytes[offset + i * 16 + 12 + j]);
                ss << str;
            }
            std::string str = ss.str();
            std::string name = GetName(cnt(offset + i * 16, 4) + sections[".strtab"]->offset);
            int value = cnt(offset + i * 16 + 4, 4);
            int size = cnt(offset + i * 16 + 8, 4);
            std::string infoStr = str.substr(0, 8);
            std::reverse(infoStr.begin(), infoStr.end());
            std::string otherStr = str.substr(8, str.size() - 8);
            std::reverse(otherStr.begin(), otherStr.end());
            int info = std::stoi(infoStr, nullptr, 2);
            int other = std::stoi(otherStr, nullptr, 2);
            std::cout << value << " " << size << " " << info << " " << other << std::endl;
            symTable[value] = elements.size();
            elements.push_back(std::make_shared<utility::symtable_element>(name, value, size, i, info, other));


        }
        RiscVParser::elements = elements;
        RiscVParser::symTable = symTable;
        RiscVParser::textAddr = sections[".text"]->addr;
    }

    void ElfParser::dumpSymTable() {
        writer->writeLine(".symtab");
        writer->writeLine("Symbol Value          \tSize Type \tBind \tVis   \tIndex Name");
        for (const auto &element: elements) {
            writer->write(element->GetString());
        }
    }

    void ElfParser::parse() {
        parseHeader();
        parseSymTab();
        parseText();
        dumpSymTable();
    }
} // parser

File: Arhitecture/AsmParser/src/ElfParser.java
import java.io.*;
import java.util.*;

public class ElfParser {
    private final short[] bytes;
    private final DataInputStream reader;
    private final BufferedWriter writer;
    private final List<SymTableEntity> entities;
    private final Map<String, Section> sections;
    private final Map<Integer, Integer> symTable;
    private int textAddr;

    public ElfParser(final String inputFileName, final String outputFileName) throws IOException {
        try {
            this.reader = new DataInputStream(new FileInputStream(inputFileName));
            this.writer = new BufferedWriter(new FileWriter(outputFileName));
            this.entities = new ArrayList<>();
            this.sections = new HashMap<>();
            this.symTable = new HashMap<>();
            byte[] tmp = reader.readAllBytes();
            this.reader.close();
            this.bytes = new short[tmp.length];

            for (int i = 0; i < bytes.length; i++) {
                bytes[i] = (short) (((short)((tmp[i] < 0) ? 256 : 0)) + ((short)tmp[i]));
            }
        } catch (IOException e) {
            System.err.println(e.getMessage());
            throw e;
        }
    }

    public void parse() throws IOException {
        parseHeader();
        parseSymTable();
        parseText();
        saveSymTable();
        writer.close();
    }

    private void parseHeader() {
        final int shoff = cnt(32, 4);
        final int shentsiz = cnt(46, 2);
        final int shnum = cnt(48, 2);
        final int shstrndx = cnt(50, 2);
        final int go = shoff + shentsiz * shstrndx;
        final int offset12 = cnt(go + 16, 4);
        for (int j = 0, i = shoff; j < shnum; j++, i += 40) {
            final String name = getName(offset12 + cnt(i, 4));
            int[] params = new int[10];
            for (int k = 0; k < 10; k++) {
                params[k] = cnt(i + 4 * k, 4);
            }
            sections.put(name, new Section(
                    params[0], params[1],
                    params[2], params[3],
                    params[4], params[5],
                    params[6], params[7],
                    params[8], params[9]
            ));
        }
    }



    private void parseSymTable() {
        final Section symtab = sections.get(".symtab");
        final int offset = symtab.offset;
        final int count = symtab.size / 16;
        for (int i = 0; i < count; i++) {
            final StringBuilder builder = new StringBuilder();
            for (int j = 0; j < 4; j++) {
                builder.append(decimalToBinary(bytes[offset  + i * 16 + 12 + j]));
            }
            final String str = builder.toString();
            final String name = getName(cnt(offset + i * 16, 4) + sections.get(".strtab").offset);
            final int value = cnt(offset + i * 16 + 4, 4);
            final int size = cnt(offset + i * 16 + 8, 4);
            final int info = Integer.parseInt(new StringBuilder(str.substring(0, 8)).reverse().toString(), 2);
            final int other = Integer.parseInt(new StringBuilder(str.substring(8)).reverse().toString(), 2);
            System.out.println(value + " " + size + " " + info + " " + other);
            symTable.put(value, entities.size());
            entities.add(new SymTableEntity(name, value, size, i, info, other));
        }
        textAddr = sections.get(".text").addr;
    }

    private void saveSymTable() throws IOException {
        writer.write(".symtab");
        writer.newLine();;
        writer.write("Symbol Value          \tSize Type \tBind \tVis   \tIndex Name");
        writer.newLine();
        for (SymTableEntity entity : entities) {
            writer.write(entity.toString());
            writer.newLine();
        }
    }

    private void parseText() {

    }

    private int cnt(int offset, int num) {
        int ans = 0;
        for (int i = num - 1; i >= 0; i--) {
            ans = ans * 256 + bytes[offset + i];
        }
        return ans;
    }

    private static String decimalToBinary(int x) {
        final StringBuilder builder = new StringBuilder(Integer.toBinaryString(x));
        while (builder.length() < 8) {
            builder.append("0");
        }
        return builder.toString();
    }

    private String getName(int offset) {
        final StringBuilder builder = new StringBuilder();
        while (bytes[offset] != 0) {
            builder.append((char)bytes[offset++]);
        }
        return builder.toString();
    }
}
